src/EndpointV2/EndpointV2Middleware.php

```diff
@@ -303,4 +311,50 @@ private function normalizeAuthScheme(array $authScheme) : array
 
         return $normalizedAuthScheme;
     }
+
+    /**
+     * This method tries to resolve an AccountId parameter from a resolved identity.
+     * We will just perform this operation if the parameter AccountId is part of the ruleset parameters and
+     * AccountIdEndpointMode is not disabled, otherwise, we will ignore it.
+     * We also are going to resolve identity from a supplied credentials provider,
+     * and we will inject the resolved identity into the $command['@context'] property, so it can be reused along the way.
+     *
+     * @param array $commandArgs
+     * @param array $rulesetParams
+     * @return void
+     */
+    private function resolveAccountIdArgument(array &$commandArgs, array $rulesetParams)
+    {
+        static $ACCOUNT_ID_PARAM = 'AccountId';
+        if (!in_array($ACCOUNT_ID_PARAM, array_keys($rulesetParams))) {
+            return;
+        }
+
+        $accountIdEndpointMode = $this->clientArgs['AccountIdEndpointMode'];
+        if ($accountIdEndpointMode === 'disabled') {
+            return;
+        }
+
+        $credentialsProviderFn = $this->credentialsProvider;
+        $identity = $credentialsProviderFn()->wait();
+        $commandArgs['@context']['resolved_identity'] = Create::promiseFor($identity);
+        if (empty($identity->getAccountId())) {
+            $messageFn = function ($mode) {
```

Comment:
The message contains too much information for a warning or exception.  I don't think this should be a local callable either- we should split this out into another method.  Use heredoc/nowdoc if you really need to, but I think the message could be contained within two < 80 char lines.  I don't think we should have a separate exception for this either.  I think we should be using a `UnresolvedEndpointException` if we hit the required case. I realize there's one place where I use an InvalidArgumentException— that should be changed too.

Reply:
We could use UnresolvedEndpointException, but throwing an exception that relate more with the issue can be some times more useful for customers, since it is more clear about what the caused the problem. I also checked other SDKs such as boto3 and they throw a specific exception for it as well. 

I agree that the message should be shorter. I am going to work on that.